### ISSUE TITLE
Performance on big files

### ISSUE THREAD

**Issue by vjeux** [prettier/prettier#4776]
I just found out that the performance metrics of prettier inside of Nuclide at Facebook:

- p50: 122ms
- p75: 203ms
- p90: 377ms

It's now awesome but totally reasonable. However, for big files like [this one (3k lines)](https://github.com/facebook/nuclide/blob/master/pkg/nuclide-vscode-language-service-rpc/lib/LspLanguageService.js) it consistently takes multiple seconds.

Now that prettier is in a good shape (not changing as often) and massively successful, it would be nice to start spending time on performance improvements. As far as I know, outside of fixing exponential complexity, we haven't spent any time trying to optimize it.

This PR is a call for help, I'm sure that there's a ton of optimizations that can be done to make prettier faster, now would be an awesome time to start working on them.

Here's the challenge, what can we do to make the following file take less time to print?
- https://github.com/facebook/nuclide/blob/master/pkg/nuclide-vscode-language-service-rpc/lib/LspLanguageService.js

Profiling what is taking time would be an awesome first step.

**Comment by sompylasar** [comment]
One more thing I tried, maybe it's already been tried and is already optimized, please tell me if it is.

`fits` and `printDocToString` uses a 3-item array to store commands in the queue. This does not take advantage of hidden classes to reuse object memory layouts.

I'm a little confused which variant should theoretically be best-performing in this case: with arrays or with classes.

> objects created by the same constructor and have the same set of properties assigned in the same order
> [...]
> use literal initializer for Arrays with mixed values
>
> – http://thlorenz.com/talks/demystifying-v8/talk.pdf

### Proposed solution

Refactor to use a constructor function that initializes the properties in a predefined order. Use `new` to create the objects.
```js
function Cmd(ind, mode, doc) {
  this.ind = ind;
  this.mode = mode;
  this.doc = doc;
}
```
```js
cmds.push(new Cmd(makeIndent(ind, options), mode, doc.contents));
```
```js
const cmds = [new Cmd(rootIndent(), MODE_BREAK, doc)];
```

**Comment by sompylasar** [comment]
### Problem

`printAstToDoc` uses `Map` for its cache keyed by nodes. 
On the test file, the stats are that it's 97.7% more misses than hits on this cache:
```js
{printAstToDocCacheHitCount: 206, printAstToDocCacheMissCount: 9083}
```

**Comment by j-f1** [comment]
This is awesome @sompylasar! 😎

A few comments:

> Memoize `getStringWidth` so that the same string is not processed multiple times.


Is there any reason you’re using an object as a map instead of a real `Map`? Is the object more performant?

> Refactor [`FastPath`] to not use `arguments` object.

Is using the spread operator (`...names`) just as bad as using `arguments`? If it’s faster, we might want to use that to preserve the original behavior of the function.

> `printAstToDoc` uses `Map` for its cache keyed by nodes.

Sounds like it’s not helping much. Remove it 🔥🔥🔥

**Comment by vjeux** [comment]
This is awesome! I can't wait to see if all those leads eventually become performance wins!

For the Map inside of printAstToDoc, unfortunately we can't get rid of it. See #2259 for context. The low hit rate is misleading as it's preventing a sub-tree evaluation.

**Comment by sompylasar** [comment]
One concern that I still have is that I could not get a good reproduction of the time measurements from subsequent runs on my machine, so I need (or someone could help out with this) to establish a way to benchmark multiple runs to get an average run time and then try to optimize that.

> Is there any reason you’re using an object as a map instead of a real Map? Is the object more performant?

I previously saw some performance benefits from native objects' index operators compared to `Map` which requires method calls. This requires more isolated measurements on this particular code base.

> Is using the spread operator (...names) just as bad as using arguments? If it’s faster, we might want to use that to preserve the original behavior of the function.

[Babel transpiles spread operator for "rest" argument to `arguments`](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABADwBQDpMEMBOBzAZwEpEBvAKEUQgQKhUQF5FdCBtABgF0BuSxYHByJUAGwCm9GE0QBGADSIYEGawLoJYPFAAWPJYgA8SiPoDUZmCVIBfcjaA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=), so it should be the same:
```js
function x(...args) {
  const x = args[0];
  for (let i = 1, ic = args.length; i < ic; ++i) {}
}
```
```js
"use strict";

function x() {
  var x = arguments.length <= 0 ? undefined : arguments[0];
  for (var i = 1, ic = arguments.length; i < ic; ++i) {}
}
```

But, I recently saw an article that voids the recommendation to not use `arguments`, among other, since Node 8+ and TurboFan. This needs confirmation from folks who knows Node internals like @bmeurer.

> For the Map inside of printAstToDoc, unfortunately we can't get rid of it. See #2259 for context. The low hit rate is misleading as it's preventing a sub-tree evaluation.

👍

**Comment by sompylasar** [comment]
Btw, the naive memoization of `getStringWidth` that I wrote will definitely leak memory as it's module-level, there needs to be a more robust LRU cache or something, and a way to scope it or reset it after the formatting is complete.

**Comment by vjeux** [comment]
The only hot place for getStringWidth is in a single function: https://github.com/prettier/prettier/blob/dcf44ffbdc2f403de02f12516b0c6d5d5813b16f/src/doc/doc-printer.js

**Comment by sompylasar** [comment]
> If you look through getStringWidth function, it calls through multiple packages that each create new intermediate strings. Most of the strings that are fed through the printer will be pure ascii, so it's a lot of overhead.

Yes, that's what I noticed, too.

> I don't believe that adding a cache is the right strategy here. I recommend we should do the following:
> - Add a fast path that checks if there are only ascii characters and if that's the case, return the length of the string

👍

**Comment by sompylasar** [comment]
BTW, for those who interested why `Object`, not `Map`, for caches: https://community.risingstack.com/the-worlds-fastest-javascript-memoization-library/
Surprisingly, in this article's benchmark, Object with null prototype has slightly worse performance than plain Object that should do prototype chain lookup. No idea why.
<img src="https://blog-assets.risingstack.com/2017/01/variable-javascript-memoization-cache.png" />

**Comment by sompylasar** [comment]
I added an option to the CLI to run a benchmark around the `format` function on the `stdin` input, here's the `Heavy (Bottom-Up)` results from `node --inspect-brk` as seen in DevTools.

`RegExp` and `replace` within `getStringWidth` seem to really be the first contender for optimization (below is the profile for 10000 repetitions).

![screen shot 2018-06-30 at 10 32 49 pm](https://user-images.githubusercontent.com/498274/42131379-b92cf4a4-7cb5-11e8-9432-1813ad58d1a0.png)

![screen shot 2018-06-30 at 10 36 27 pm](https://user-images.githubusercontent.com/498274/42131397-217e5156-7cb6-11e8-8b90-b0a8f1dad28e.png)

Also, probably caused by V8 runtime optimization and/or CPU caches and everything that helps processing the same code and data faster, the average time is down to 181ms on 10000 repetitions, 189ms on 100.
```
{"benchmark":"1000","totalMs":181064.38481092453,"averageMs":181.06438481092454}
```
```
{"benchmark":"100","totalMs":18986.71142411232,"averageMs":189.8671142411232}
```

Another observation is that `printGenerically` is reported as `Not optimized: Deoptimized too many times` (below is the profile for 100 repetitions):
![screen shot 2018-06-30 at 10 41 25 pm](https://user-images.githubusercontent.com/498274/42131434-c191e252-7cb6-11e8-9720-c579d96e1689.png)

**Comment by sompylasar** [comment]
Below is the profile for 10 repetitions:
![screen shot 2018-06-30 at 10 47 36 pm](https://user-images.githubusercontent.com/498274/42131478-aae05f56-7cb7-11e8-9760-fe862982d195.png)

```
$ cat LspLanguageService.js | NODE_ENV=production node ./bin/prettier.js --stdin-filepath LspLanguageService.js --benchmark 1 > /dev/null
{"benchmark":"1","totalMs":461.6846880912781,"averageMs":461.6846880912781}

$ cat LspLanguageService.js | NODE_ENV=production node ./bin/prettier.js --stdin-filepath LspLanguageService.js --benchmark 10 > /dev/null
{"benchmark":"10","totalMs":2507.7312729358673,"averageMs":250.77312729358673}

$ cat LspLanguageService.js | NODE_ENV=production node ./bin/prettier.js --stdin-filepath LspLanguageService.js --benchmark 100 > /dev/null
{"benchmark":"100","totalMs":17653.580070972443,"averageMs":176.5358007097244}

$ cat LspLanguageService.js | NODE_ENV=production node ./bin/prettier.js --stdin-filepath LspLanguageService.js --benchmark 1000 > /dev/null
{"benchmark":"1000","totalMs":164849.50500822067,"averageMs":164.84950500822066}
```

The timing displayed is for `format` function only, not the full startup time.
```js
function formatStdin(context) {
  // ...

      let formatted;
      const benchmark = context.argv["benchmark"];
      if (benchmark > 0) {
        const performance = require("perf_hooks").performance;
        const startMs = performance.now();
        for (let i = 0; i < benchmark; ++i) {
          formatted = format(context, input, options);
        }
        const totalMs = performance.now() - startMs;
        process.stderr.write(
          JSON.stringify({
            benchmark: benchmark,
            totalMs: totalMs,
            averageMs: totalMs / benchmark
          }) + "\n"
        );
      } else {
        formatted = format(context, input, options);
      }

      writeOutput(context, formatted, options);
```

**Comment by sompylasar** [comment]
@vjeux I gave your recommendation about `getStringWidth` another thought.

> I don't believe that adding a cache is the right strategy here. I recommend we should do the following:
> - Add a fast path that checks if there are only ascii characters and if that's the case, return the length of the string

This fast path that checks is still a full scan on every string that comes in (possibly with a `RegExp`) to only discover that nothing should have been done with the string. The cache, on the other hand, should be a lookup (`O(1)` if the measured string is interned) that gives the previously calculated length right away regardless of the contents of the string.

What do you think?

**Comment by sompylasar** [comment]
> Most strings are tiny, O complexity doesn’t really apply here. Scanning a string to see if all characters are ascii is one of the fastest operation a cpu can do. Faster than allocating memory to put in a hash map and doing a lookup.

Yes, I hear you. Although I believe it should not copy the string into the hashmap if the key string is already in the JS heap (but no hard evidence of that).

> But I don’t really trust my instinct on those things; just build it and see if it changes anything in the perf.

Yes, that's why I started from measuring, and then tweaked the code to find opportunities to optimize.

> For your comment about startup time, we keep the node process alive so I’m not super interested in this.

Good to know, this means [the benchmark options I added](https://github.com/sompylasar/prettier/commit/3ececd399e91189dfe59f830913ff9e72b8c5271) are relevant to measure just the function you're interested in optimizing.

Please have a look (tell me if you want to see it in a PR, and should the PR include the benchmark options commit):
https://github.com/sompylasar/prettier/commit/d43e0221fcccf08c7ea77886ced880b7556356ff
```js
// eslint-disable-next-line no-control-regex
const notAsciiRegex = /[^\x20-\x7F]/;

function getStringWidth(text) {
  if (!text) {
    return 0;
  }

  // shortcut to avoid needless string `RegExp`s, replacements, and allocations within `string-width`
  if (!notAsciiRegex.test(text)) {
    return text.length;
  }

  // emojis are considered 2-char width for consistency
  // see https://github.com/sindresorhus/string-width/issues/11
  // for the reason why not implemented in `string-width`
  return stringWidth(text.replace(emojiRegex, "  "));
}
```

`getStringWidth` dropped significantly in the Profiler `Heavy (Bottom-Up)`.

### Before
Total run time for 1000 repetitions for `LspLanguageService.js` is **182276ms**, average run is **181.80ms**.
```
$ cat LspLanguageService.js | NODE_ENV=production node --inspect-brk ./bin/prettier.js --stdin-filepath LspLanguageService.js --loglevel debug --debug-repeat 1000 > /dev/null
Debugger listening on ws://127.0.0.1:9229/d085f49d-a12c-432e-ad7e-acf322152bd3
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
[debug] normalized argv: {"color":true,"editorconfig":true,"stdin-filepath":"LspLanguageService.js","loglevel":"debug","debug-repeat":"1000","plugin-search-dir":[],"plugin":[],"ignore-path":".prettierignore","config-precedence":"cli-override","_":[]}
[debug] resolve config from '/Users/ivanbabak/_sompylasar/_github/prettier/LspLanguageService.js'
[debug] loaded options `null`
[debug] applied config-precedence (cli-override): {"filepath":"LspLanguageService.js"}
[debug] '--debug-repeat' option found, running formatWithCursor 1000 times.
[debug] '--debug-repeat' measurements for formatWithCursor: {
[debug]   "repeat": 1000,
[debug]   "hz": 5.500298017789999,
[debug]   "ms": 181.80833052420616
[debug] }
Waiting for the debugger to disconnect...
```
![before](https://user-images.githubusercontent.com/498274/42139686-8acd1c8c-7d47-11e8-9256-c9e3110e2545.png)


### After
Total run time for 1000 repetitions for `LspLanguageService.js` is **145246ms**, average run is **144.85ms (20.32% speed-up)**.
```
$ cat LspLanguageService.js | NODE_ENV=production node --inspect-brk ./bin/prettier.js --stdin-filepath LspLanguageService.js --loglevel debug --debug-repeat 1000 > /dev/null
Debugger listening on ws://127.0.0.1:9229/1a338c30-c52c-4ae7-8e11-9aac0a2b736a
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
[debug] normalized argv: {"color":true,"editorconfig":true,"stdin-filepath":"LspLanguageService.js","loglevel":"debug","debug-repeat":"1000","plugin-search-dir":[],"plugin":[],"ignore-path":".prettierignore","config-precedence":"cli-override","_":[]}
[debug] resolve config from '/Users/ivanbabak/_sompylasar/_github/prettier/LspLanguageService.js'
[debug] loaded options `null`
[debug] applied config-precedence (cli-override): {"filepath":"LspLanguageService.js"}
[debug] '--debug-repeat' option found, running formatWithCursor 1000 times.
[debug] '--debug-repeat' measurements for formatWithCursor: {
[debug]   "repeat": 1000,
[debug]   "hz": 6.903349947385214,
[debug]   "ms": 144.85720811223985
[debug] }
Waiting for the debugger to disconnect...
```
![after-1](https://user-images.githubusercontent.com/498274/42139676-6107c5dc-7d47-11e8-9a32-5b9f446b2386.png)
![after-2](https://user-images.githubusercontent.com/498274/42139622-9560c870-7d46-11e8-88fe-7eed316d0cfc.png)

**Comment by vjeux** [comment]
Wow! 20% faster by adding a trivial test :) Ship it!

**Comment by sompylasar** [comment]
@vjeux 👉 https://github.com/prettier/prettier/pull/4788 (both the CLI options and the getStringWidth perf improvement)

**Comment by sompylasar** [comment]
Okay, I take [this](https://github.com/prettier/prettier/issues/4776#issuecomment-401709194) back – bundled version starts having this `if (isNotProduction) {` statement if I use a variable. 😞

**Comment by j-f1** [comment]
Also, (IMO) the performance of the unbundled/development versions shouldn’t be as important as the performance of the bundled/production version.

**Comment by sompylasar** [comment]
> Also, (IMO) the performance of the unbundled/development versions shouldn’t be as important as the performance of the bundled/production version.

Agree, but for the sake of quick iterations on the code, it's nice to have at least NODE_ENV=production runs not much worse than the bundled version runs. The build takes quite some time, also now there's no way known to me to build just one variant. Of course, it's more correct to measure the build, not the source as I do, as there's an optimizing compiler in place (rollup). Maybe someone could split and speed up the build to improve development iteration times.

**Comment by duailibe** [comment]
@sompylasar the build system has a cache now. It should only recompile an artifact if:
- one of the source files included in the artifact was changed;
- yarn.lock changed (which will rebuild everything)
- or you bump the version prefix.

We probably need to update CONTRIBUTING.md

**Comment by sompylasar** [comment]
@duailibe "one of the source files included in the artifact was changed;" – right, and this is what I'm doing by working on this ticket, changing the source files. 😄

**Comment by duailibe** [comment]
@sompylasar I'm aware, but you're likely to change files that will only affect `index.js`, `bin-prettier.js` and `standalone.js` which build reasonably fast. The parsers (specially Flow and TypeScript) are the culprits of taking too long, so you should be fine :P

**Comment by sompylasar** [comment]
New improvement 👉 https://github.com/prettier/prettier/pull/4848

**Comment by sompylasar** [comment]
New npm scripts and documentation for performance benchmarks 👉   https://github.com/prettier/prettier/pull/4846

**Comment by sanmai-NL** [comment]
@vjeux What's the status of this work in 2020? Prettier is still very slow for me on big files. Multiple seconds per formatting round.

**Comment by vjeux** [comment]
@sanmai-NL would you be able to share this file by any chance? It may be an issue for a specific pattern in your file that we may be able to address.

**Comment by sanmai-NL** [comment]
@vjeux https://han-aim.gitlab.io/dt-sd-asd/materials/ADP/ the source, and I suppose the even larger rendered HTML as well.

**Comment by fisker** [comment]
I had a quick look, this html is slow because of https://github.com/prettier/prettier/blob/19a8df2f572f1bc4504cebe0614f45f868b786fa/src/language-html/print-preprocess.js#L32, we should optimize this function, it map ast 10 times, and each map cost about 1s.

**Comment by jpike88** [comment]
Rome has been released with formatting performance many times more than Prettier:
https://rome.tools/blog/2022/04/05/rome-formatter-release?ref=console.dev

Maybe it's time to make performance more of a priority

**Comment by sdwvit** [comment]
Formatting 100 mb json with prettier takes 100+ seconds and 16GB of ram, while `fs.writeFileSync(..., JSON.stringify(json, null, 2))` takes only 2 seconds and 700mb of ram... 
I know it is only one use case, but I expect prettier to be no slower than plain JSON.stringify.
